program/src/balance.rs

```diff
@@ -102,6 +102,63 @@ pub fn get_target_balance(
     Ok(target_balance)
 }
 
+/// Get the index of the validator to unstake from, if we need to unstake at all.
+
+/// If any validator is more than threshold away from its target, this function
+/// will try to unstake, and return the index of the validator where unstaking
+/// will have the largest impact.
+pub fn get_unstake_validator_index(
+    validators: &Validators,
+    target_balance: &[Lamports],
+    threshold: Rational,
+) -> Option<(usize, Lamports)> {
+    // Check if we need to rebalance because a validator is too far away from
+    // its target.
+    let needs_unstake =
+        validators
+            .entries
+            .iter()
+            .zip(target_balance.iter())
+            .any(|(validator, target)| {
+                let target_difference = target
+                    .0
+                    .saturating_sub(validator.entry.effective_stake_balance().0);
+                if target == &Lamports(0) {
+                    return false;
+                }
+                Rational {
+                    numerator: target_difference,
+                    denominator: target.0,
+                } >= threshold
+            });
+
+    let (idx, validator) = validators
+        .entries
+        .iter()
+        .enumerate()
```

Comment:
We could zip it with `target_balance` and avoid the indexing:

```rust
validators
    .entries
    .iter()
    .zip(&target_balance)
    .max_by_key(|(validator, target)| ...)
```

Reply:
Yeah, but we still need to return the index if we want to later get the stake account associated with the validator. We could pass it on the function, but I think it's not necessary.

Reply:
Fair point, never mind then.